qiskit/pulse/timeslots.py

```diff
@@ -193,68 +270,163 @@ def duration(self) -> int:
+        if channel in self._table:
+            return tuple(self._table[channel])
+
+        return tuple()
+
+    def ch_start_time(self, *channels: Channel) -> int:
         """Return earliest start time in this collection.
 
         Args:
             *channels: Channels over which to obtain start_time.
         """
-        intervals = list(itertools.chain(*(self._table[chan] for chan in channels
+        timeslots = list(itertools.chain(*(self._table[chan] for chan in channels
                                            if chan in self._table)))
-        if intervals:
-            return min(interval.begin for interval in intervals)
+        if timeslots:
+            return min(timeslot.start for timeslot in timeslots)
+
         return 0
 
-    def ch_stop_time(self, *channels: List[Channel]) -> int:
+    def ch_stop_time(self, *channels: Channel) -> int:
         """Return maximum time of timeslots over all channels.
 
         Args:
             *channels: Channels over which to obtain stop time.
         """
-        intervals = list(itertools.chain(*(self._table[chan] for chan in channels
+        timeslots = list(itertools.chain(*(self._table[chan] for chan in channels
                                            if chan in self._table)))
-        if intervals:
-            return max(interval.end for interval in intervals)
+        if timeslots:
+            return max(timeslot.stop for timeslot in timeslots)
+
         return 0
 
-    def ch_duration(self, *channels: List[Channel]) -> int:
+    def ch_duration(self, *channels: Channel) -> int:
         """Return maximum duration of timeslots over all channels.
 
         Args:
             *channels: Channels over which to obtain the duration.
         """
         return self.ch_stop_time(*channels)
 
-    def is_mergeable_with(self, timeslots: 'TimeslotCollection') -> bool:
+    def is_mergeable_with(self, timeslot_collection: 'TimeslotCollection') -> bool:
         """Return if self is mergeable with `timeslots`.
 
         Args:
-            timeslots: TimeslotCollection to be checked
+            timeslot_collection: TimeslotCollection to be checked for mergeability
         """
-        for slot in timeslots.timeslots:
-            if slot.channel in self.channels:
-                for interval in self._table[slot.channel]:
-                    if slot.interval.has_overlap(interval):
+        common_channels = set(self.channels) & set(timeslot_collection.channels)
+
+        for channel in common_channels:
+            ch_timeslots = self._table[channel]
+            other_ch_timeslots = timeslot_collection._table[channel]
+
+            for other_ch_timeslot in other_ch_timeslots:
+                other_ch_interval = other_ch_timeslot.interval
+
+                append = True
+                for ch_timeslot in ch_timeslots:
+                    ch_interval = ch_timeslot.interval
+                    if other_ch_interval > ch_interval:
```

Comment:
I think you should be comparing the first interval in `other_ch_timeslots` to the _last_ interval in `ch_timeslots` if you're checking for append! Both lists of timeslots should be sorted right?

Reply:
😭 This function is hard to understand.

Can you separate the checking for append out? You can also make it worst case linear time if we know the timeslots are indeed sorted.
```
for channel in common_channels:
    if ch_timeslots[-1].stop < other_ch_timeslots[0].start:
        continue  # We are appending along this channel
    
    i = 0  # iterate through this
    j = 0  # iterate through other
    while i < len(ch_timeslots) and j < len(other_ch_timeslots):
        if ch_timeslots[i].interval.has_overlap(other_ch_timeslots[j].interval):
            return False
        if ch_timeslots[i].stop <= other_ch_timeslots[j].start:
            i += 1
        else:
            j += 1
return True
```

Reply:
> I think you should be comparing the first interval in other_ch_timeslots to the last interval in ch_timeslots if you're checking for append! Both lists of timeslots should be sorted right?

yes, good catch. I missed this one.